### PR TITLE
allows whitelisted symlinks to be imported from

### PR DESCRIPTION
**Pull request by bwlang**
needs testing, but I can't quickly find where the other methods in remote_files are tested...  
@natefoo can you comment on the implementation and you point me in the right direction on testing?

### CODE REVIEW

File: lib/galaxy/util/path/__init__.py [hunk 1]
```diff
@@ -33,6 +33,17 @@
 from six.moves import map, zip
 log = logging.getLogger(__name__)
 
+def path_under_whitelist_dir(path, whitelist=None):
```
Comment (natefoo) [inline lib/galaxy/util/path/__init__.py]: I think this duplicates `safe_contains()` below? You should be able to use `safe_contains(configured import root directory, path, whitelist)` I think.

File: lib/galaxy/webapps/galaxy/api/remote_files.py [hunk 2]
```diff
@@ -123,15 +126,16 @@ def __load_all_filenames(self, directory):
...
         and its subfolders and returns jstree representation
         of its structure.
         """
-        userdir_jstree = None
         jstree_paths = []
-        if os.path.exists(directory) and not os.path.islink(directory):
+        if os.path.exists(directory):
+            if os.path.islink(directory) and not self.__safe_directory(directory, whitelist):
+                raise exceptions.ConfigDoesNotAllowException('directory (%s) is a symlink to a location not on the whitelist' % directory)
             for (dirpath, dirnames, filenames) in os.walk(directory):
```
Comment (natefoo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: I think the above would not be needed if this used `galaxy.util.path.unsafe_walk()` with the whitelist.

**Review by natefoo — CHANGES_REQUESTED** [review]
Thanks @bwlang! I think there is some duplication here, although I might be missing something.

**Comment by bwlang** [comment]
@natefoo 
I thought you'd want to allow symlinks pointing anywhere that's whitelisted, not just directories below the import root directory. Is that right?

where should the testing live for this code?

**Comment by natefoo** [comment]
Ok, `safe_contains(directory, directory, whitelist)` does what you want, but feels a little clumsy, so your new function seems reasonable. I have a `safe_walk()` function I'd like to add, though. Do you mind if I PR against your branch?

**Comment by bwlang** [comment]
Looking better after Nate's contribution... still not sure about testing this though.

**Comment by natefoo** [comment]
I'm told by @martenson that this failing test is flaky.

**Comment by natefoo** [comment]
@galaxybot test this

**Comment by bwlang** [comment]
That's some chatty selenium output...
maybe this test is not robust... seems like a timeout error unrelated to this commit to my eye.
@natefoo  thoughts?
```
======================================================================
ERROR: test_build_paired_list_hide_original (selenium_tests.test_collection_builders.CollectionBuildersTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/galaxy/test/selenium_tests/framework.py", line 148, in func_wrapper
    return f(self, *args, **kwds)
  File "/galaxy/test/selenium_tests/test_collection_builders.py", line 79, in test_build_paired_list_hide_original
    self.history_panel_wait_for_hid_visible(2)
  File "/galaxy/test/galaxy_selenium/navigates_galaxy.py", line 267, in history_panel_wait_for_hid_visible
    self.wait(timeout).until(history_has_hid)
  File "/galaxy_venv/local/lib/python2.7/site-packages/selenium/webdriver/support/wait.py", line 80, in until
    raise TimeoutException(message, screen, stacktrace)
TimeoutException: Message: 
```

**Comment by natefoo** [comment]
Yeah I don't think we broke this - someone else should perhaps +1/merge this since I contributed to it.

**Comment by martenson** [comment]
Thanks @bwlang, @natefoo !

**Pull request merged**